### PR TITLE
Draft/compiler flags

### PR DESCRIPTION
**Pull request by serge-sans-paille**

### CODE REVIEW

File: content/2016-03-29-compiler-flags.rst [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+compiler than only the functions flagged with a special attribute are part of
+the external ABI, the other ones are not exported. As Python uses a single
+entry point to load Pythran modules, namely ``PyInit_cda`` for Python3 modules
+and ``initcda`` for Python2 modules [1]_, one can add the ``__attribute__
+((visibility("default")))`` on this symbol and it will be the only exported
+one. This slightly impacts the code size and eventually gives the compiler more
+optimization opportunities, but nothing significant there (131K), apart the pleasure
+of generating cleaner binaries. That's also going to be the default for next
+Pythran version.
+
```
Comment (aguinet) [inline content/2016-03-29-compiler-flags.rst:144]: You can also say here that carefully choosing which symbols are exported can improve loading time among othet things, as cited here : https://gcc.gnu.org/wiki/Visibility

File: content/2016-03-29-compiler-flags.rst [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+Out of chance: getting faster binaries
+======================================
+
+In the (huge) info pages of GCC, near the doc of ``-fvisibility=hidden``,
+there's this (GCC only) compiler flag, ``-fwhole-program`` that implements some
+kind of Link Time Optimization, in the sense that it tells the compiler to
+consider the current compilation unit (or code) as a whole program. ``main`` is
```
Comment (aguinet) [inline content/2016-03-29-compiler-flags.rst]: Suggestion en remplacement de "`main` is ..." : As specified in the GCC man page, "All public functions and variables with the exception of "main" and those merged by attribute "externally_visible" become static functions and in effect are optimized more aggressively by interprocedural optimizers.", which basically means that every function is considered static except for "main" and the ones that are explicitly told not to be. This allows the compiler for instance to remove functions that are always inlined, and thus win space.

File: content/2016-03-29-compiler-flags.rst [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+    1000 loops, best of 3: 1.02 msec per loop
+
+If you're really into compiler flags tuning, you can try out ``-funroll-loops``
+or try to tune the ``-finline-limit=N`` parameter (that actually get mets dow
+to ``1ms per loop``) but that's going a bit too far :-)
+
+Don't forget Vectorization
+==========================
+
+Combining ``-O3`` and ``-march=native`` triggers compiler auto-vectorization,
```
Comment (aguinet) [inline content/2016-03-29-compiler-flags.rst]: only GCC needs this, clang turns vectorisation at -O2 (SSE2 IIRC w/ x86). -march=native will allow him to use a more recent instruction set if available

File: content/2016-03-29-compiler-flags.rst [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+        phi1 = (90.0 - lat1)*degrees_to_radians
+        phi2 = (90.0 - latitudes)*degrees_to_radians
+        theta1 = long1*degrees_to_radians
+        theta2 = longitudes*degrees_to_radians
+        cos = (math.sin(phi1)*np.sin(phi2)*np.cos(theta1 - theta2) +
+               math.cos(phi1)*np.cos(phi2))
+        arc = np.arccos( cos )
+        return np.argmin(arc), arc.min()
+
+It doesn't even wight a kilobyte, and when benchmarked, it runs in a few milliseconds:
```
Comment (pbrunet) [inline content/2016-03-29-compiler-flags.rst]: wight?

Comment (serge-sans-paille) [inline content/2016-03-29-compiler-flags.rst]: > wight?
> weight

**Pull request merged**